lib/task_manager.js

```diff
@@ -0,0 +1,67 @@
+class TaskManager {
+  constructor() {
+    this.map = {};
+    this.queue;
+    this.buffer = [];
+    this.lastPushedToQueue = 0;
+  }
+
+  async initQueue(maxConcurrentRequests) {
+    const PQueue = (await import('p-queue')).default;
+    this.queue = new PQueue({ concurrency: maxConcurrentRequests });
+    this.queue.on('completed', (data) => this.handleNewData(data));
+    // this.queue.on('error', err => { return Promise.reject(err.toString()); } );
+  }
+
+  clearQueue() {
+    this.taskManager.queue.pause();
```

Comment:
pause doesn't pause already in flight tasks. We clean queue next step which won't work if there's a long running task running